### PR TITLE
Ongoing Scala 3 support

### PR DESCRIPTION
**Pull request by jrudolph**
Continuation of #274

This PR documents the ongoing progress on Scala 3 support. It stays open until the support is ready for merging into master.

The commits are supposed to stay, so let's try to use reasonable commits. Changes can be PR'd against this branch.

- [x] sbt configuration
- [x] [core] structure for compilation without any features
- [x] [core] implementation of macros in scala 3
	- [x] Top-level `run` macro @lrytz #289 
	- [x] CombinatorSpec (#287)
	- [x] ActionSpec
      - [x] `run` combinator #295 
      - [x] RuleCall remaining cases
      - [x] `test` @yanns #291
      - [x] `~>` remaining cases
      - [x] Try to revert changes in Calculator2 examples if case class construction can be fixed
    - [x] ErrorReportingSpec fails (also one test in CalculatorSpec fails because of error reporting) #303
    - [x] CutSpec #292
- [x] Figure remaining type-related issues
	- [x] Some commented tests in ActionSpec either fail to type or even get into loop
- [x] [core] all tests passed on scala 2 and scala 3
- [x] port DynamicRuleDispatch macro #299
- [x] `parboiled` module tests pass #299
- [x] `example` module tests pass #298
- [x] Optimizations (i.e. works correctly without but some clever translations are still missing)
  - [x] `times` for statically known boundary values
  - [x]  rules based on statically constant Strings #294
- [x] all tests passed on scala 2 and scala 3

### CODE REVIEW

**Comment by jrudolph** [comment]
@nicolasstucki pattern matching calls to this method with quoted patterns seems quite hard because of the complex relationship of all the type parameters involved:

https://github.com/sirthias/parboiled2/blob/master/parboiled-core/src/main/scala/org/parboiled2/RuleDSLActions.scala#L30

It seems easy enough to parse them using trees/terms but is there any chance of matching them using quoted patterns?

E.g. the first problem is that you cannot even put in wildcard types for the type parameters because they don't fulfill the bounds:

```scala
case '{($p: Parser).capture[a, b]($arg)($d)} =>
```

fails with

```
[error]     |Type argument a does not conform to upper bound org.parboiled2.support.hlist.HList in subpart org.parboiled2.Rule[a, b] of inferred type scala.quoted.Expr[org.parboiled2.Rule[a, b]]
```

Should it work leaving off the type arguments completely? In that case, the macro compiles but it does not match the invocation when the macro is executed. Are there any extra tools to debug non-matching quote patterns?

**Comment by nicolasstucki** [comment]
If I understand correctly you need to add upper bounds to `a` and `b`. If that is the case then you can use

```scala
case '{ type a <: T; type b < T2; ($p: Parser).capture[`a`, `b`]($arg)($d)} =>
```

**Comment by jrudolph** [comment]
Cool, I'm going to try that.

**Comment by jrudolph** [comment]
> If I understand correctly you need to add upper bounds to `a` and `b`. If that is the case then you can use
> 
> ```scala
> case '{ type a <: T; type b < T2; ($p: Parser).capture[`a`, `b`]($arg)($d)} =>
> ```

Something similar compiles but still doesn't match:

```scala
case '{ type I <: HList; type O <: HList; ($p: Parser).capture[I, O]($arg: Rule[I, O])($d) } =>
```

doesn't match this tree:

```
$anon.this.capture[org.parboiled2.support.hlist.HNil, org.parboiled2.support.hlist.HNil]($anon.this.ch('b'))(org.parboiled2.support.hlist.ops.hlist.Prepend.hnilPrepend1[org.parboiled2.support.hlist.HNil, org.parboiled2.support.hlist.::[scala.Predef.String, org.parboiled2.support.hlist.HNil]])
```

**Comment by jrudolph** [comment]
> If I understand correctly you need to add upper bounds to `a` and `b`. If that is the case then you can use
> 
> ```scala
> case '{ type a <: T; type b < T2; ($p: Parser).capture[`a`, `b`]($arg)($d)} =>
> ```

Something similar compiles but still doesn't match:

```scala
case '{ type I <: HList; type O <: HList; ($p: Parser).capture[I, O]($arg: Rule[I, O])($d) } =>
```

doesn't match this tree:

```
$anon.this.capture[org.parboiled2.support.hlist.HNil, org.parboiled2.support.hlist.HNil]($anon.this.ch('b'))(org.parboiled2.support.hlist.ops.hlist.Prepend.hnilPrepend1[org.parboiled2.support.hlist.HNil, org.parboiled2.support.hlist.::[scala.Predef.String, org.parboiled2.support.hlist.HNil]])
```

**Comment by jrudolph** [comment]
I updated the task list above. Let's put our names behind items if we start working on it to avoid stepping on each others' toes.

**Comment by lrytz** [comment]
> Let's put our names

I don't have permission, I'm on the `run` macro

**Comment by yanns** [comment]
I start `test()` (tested in `ActionSpec`)

**Comment by jrudolph** [comment]
I'll also take the `run` combinator which seems to be at the heart of the remaining typing issues.

**Comment by sirthias** [comment]
Not to distract you guys from the current commit frenzy here, but I still would like to applaud everyone from the sidelines for the amazing progress you are making on the Scala 3 port.
Normally I would definitely take a day and help out on things wherever possible but I'm going to be on family vacation for another week, so a little cheering is all I can do at this point...
Aynway: ❤🥳🙏👏🤞

**Comment by yanns** [comment]
`CutSpec` in https://github.com/sirthias/parboiled2/pull/292

**Comment by jrudolph** [comment]
Thanks, Mathias!

FYI, not working today and will only continue tomorrow.

**Comment by jrudolph** [comment]
> case '{ type I <: HList; type O <: HList; ($p: Parser).capture[I, O]($arg: Rule[I, O])($d) } =>

Ok, that's not what you wrote, @nicolasstucki. If I do it exactly as you pointed out, it works indeed:

```
case '{ type i <: HList; type o <: HList; ($p: Parser).capture[`i`, `o`]($arg)($l) } => 
```

That's great and will simplify the code a lot!

**Comment by jrudolph** [comment]
Found a quirk in the Scala compiler when converting the `RunResult` implicit logic to match types: https://github.com/lampepfl/dotty/issues/13455.

**Comment by jrudolph** [comment]
Turns out that `RunResult` probably cannot be implement using a type match any ways because it's impossible to match on open type hierarchies. This means you cannot distinguish between function types like `T1 => R` and `(T1, T2) => R`. Let's see if the existing implicit based structure works good enough for these types.

**Comment by jrudolph** [comment]
Unfortunately, the heavy-machinery type-based stuff is hard to fix because it's so hard to understand why it doesn't work. I get pages-long error messages about why `TailSwitch` couldn't be reduced, which look as if in some cases the compiler gave up somewhere and replaced a concrete HList type by just `HList`.

So, it seems that the type-match based `TailSwitch` does not work well (the former implicit version is also broken). At least in combination with the Action typing it goes off completely and gets into an unlimited loop. If someone want to have a try, you can look at the tests still commented in ActionSpec.

**Comment by sirthias** [comment]
> ... (the former implicit version is also broken).

Maybe this [long-standing bug here](https://github.com/sirthias/parboiled2/issues/172) is somehow related.
The old implicit-based `TailSwitch` logic is indeed hard to read/understand and apparently not without issues.
Computing the type of parboiled2 rule in the general case it was definitely the most complicated type machinery I've ever written. It would be great to see this simplified with the new language features...

**Comment by jrudolph** [comment]
My suspicion is either that expected types lead to some issues in Scala 3, or it's that `TailSwitch` has multiple outputs which might get in the way of inference.

Since I've not too much time left on this, I might try removing the generic reduction-rule mechanism to simplify things (i.e. completely remove `run` and the `I` type parameter of `Rule`. @yanns, do you use `PopRule`s? I.e. rules with the input parameter not being `HNil`?

@sirthias it seems there's some overlap in functionality between `run` and `~>`. I guess `run` needs those pop rules because these rules are standalone and only then get combined but `~>` might probably work without reduction in usual grammars. IIRC reduction rules were quite infrequent in the grammars we used to build.

**Comment by jrudolph** [comment]
> Maybe this [long-standing bug here](https://github.com/sirthias/parboiled2/issues/172) is somehow related.

Oh yes, contravariance...

**Comment by yanns** [comment]
> Since I've not too much time left on this, I might try removing the generic reduction-rule mechanism to simplify things (i.e. completely remove `run` and the `I` type parameter of `Rule`. @yanns, do you use `PopRule`s? I.e. rules with the input parameter not being `HNil`?

Sorry, I have not had much time to work on this in the last days.
I don't really remember having rules like those. But I'm not 100% sure right now.

**Comment by yanns** [comment]
I've tried to find out why the test in CalculatorSpec is failing because of error reporting.

```
def InputLine = rule (Expression ~ EOI)
```
is currently parsed into `RuleCall`.
In the error message, `InputLine` does not appear. Only `EOI` is mentioned. So it means that we should not parse this into `RuleCall` but into something else that is transparent for the rule traces.

**Comment by jrudolph** [comment]
> In the error message, `InputLine` does not appear. Only `EOI` is mentioned. So it means that we should not parse this into `RuleCall` but into something else that is transparent for the rule traces.

Hmm, that sounds like the functionality of `ctx.topLevel`. Do you think something is wrong about that? Maybe we can compare generated code between Scala 2 and 3 to see what might be wrong?

**Comment by sirthias** [comment]
One question here:
How do you guys debug your current work?
I've used something like https://github.com/sirthias/macrolizer in the past to help me see, how a macro expands at a specific call site. Is there something similar already and easily available for Scala3?
(Its very easy to build oneself, of course. I'm simply interested in your process here...)

**Comment by yanns** [comment]
> One question here:
> How do you guys debug your current work?
> I've used something like https://github.com/sirthias/macrolizer in the past to help me see, how a macro expands at a specific call site. Is there something similar already and easily available for Scala3?
> (Its very easy to build oneself, of course. I'm simply interested in your process here...)

Pretty artisanal for me: 
- adding some println like `println(outerRule.asTerm.show(using Printer.TreeStructure))`.
- compile everything
- change one test just to re-compile that part and check the output.

**Comment by jrudolph** [comment]
I was using `set scalacOptions += "-Xprint:staging` but I actually didn't need it often.

**Comment by yanns** [comment]
> > In the error message, `InputLine` does not appear. Only `EOI` is mentioned. So it means that we should not parse this into `RuleCall` but into something else that is transparent for the rule traces.
> 
> Hmm, that sounds like the functionality of `ctx.topLevel`. Do you think something is wrong about that? Maybe we can compare generated code between Scala 2 and 3 to see what might be wrong?

I think that `ctx.topLevel` is all right, as it is using a `Left(...)` that is transparent for the rule traces. The issue is about a `RuleCall(Right(...))`. I'll dig deeper, comparing with scala 2.

**Comment by lrytz** [comment]
I was also using `Compile / console / scalacOptions += "-Xprint:inlining",` then `parboiledCoreJVM/console` and then write some parser class in the REPL

**Comment by sirthias** [comment]
@jrudolph @yanns @lrytz
Interesting! Thank you.

**Comment by yanns** [comment]
Oh we might have broken one test in scala 2.13:
```
[error] /Users/yannsimon/projects/parboiled2/parboiled-core/src/test/scala/org/parboiled2/RunningSpec.scala:27:37: Invalid int base expression for `.times(...)`: n.-(m)
[error]     def C(n: Int)(m: Int) = rule((n - m).times('c'))
[error]                                     ^
```

**Comment by yanns** [comment]
> > > In the error message, `InputLine` does not appear. Only `EOI` is mentioned. So it means that we should not parse this into `RuleCall` but into something else that is transparent for the rule traces.
> > 
> > 
> > Hmm, that sounds like the functionality of `ctx.topLevel`. Do you think something is wrong about that? Maybe we can compare generated code between Scala 2 and 3 to see what might be wrong?
> 
> I think that `ctx.topLevel` is all right, as it is using a `Left(...)` that is transparent for the rule traces. The issue is about a `RuleCall(Right(...))`. I'll dig deeper, comparing with scala 2.

In scala 2, we also have:
```
RuleCall(Right(Calculator.this.Expression),"Expression")
CharMatch(Calculator.this.EOI)
```

Wrong assumption. Back to investigation start :D

**Comment by jrudolph** [comment]
> Oh we might have broken one test in scala 2.13:

That's a new variant of the test that @lrytz added for what the Scala 2 macro seems to be too strict. Fixing...

**Comment by yanns** [comment]
> > > > In the error message, `InputLine` does not appear. Only `EOI` is mentioned. So it means that we should not parse this into `RuleCall` but into something else that is transparent for the rule traces.
> > > 
> > > 
> > > Hmm, that sounds like the functionality of `ctx.topLevel`. Do you think something is wrong about that? Maybe we can compare generated code between Scala 2 and 3 to see what might be wrong?
> > 
> > 
> > I think that `ctx.topLevel` is all right, as it is using a `Left(...)` that is transparent for the rule traces. The issue is about a `RuleCall(Right(...))`. I'll dig deeper, comparing with scala 2.
> 
> In scala 2, we also have:
> 
> ```
> RuleCall(Right(Calculator.this.Expression),"Expression")
> CharMatch(Calculator.this.EOI)
> ```
> 
> Wrong assumption. Back to investigation start :D

the traces are a bit different:
2.13:
```
RuleTrace(List(NonTerminal(Named(InputLine),-3), NonTerminal(RuleCall,-3), NonTerminal(Named(Expression),-3), NonTerminal(RuleCall,-3)
```

3.x
```
RuleTrace(List(NonTerminal(Named(InputLine),-3), NonTerminal(RuleCall,-3), NonTerminal(Sequence,-3), NonTerminal(Named(Expression),-3), NonTerminal(RuleCall,-3)
```
There's a `NonTerminal(Named(Expression),-3)` more. 🕵️

**Comment by jrudolph** [comment]
> the traces are a bit different:

I found it, `Sequence` was missing a bit.

**Comment by yanns** [comment]
> > the traces are a bit different:
> 
> I found it, `Sequence` was not yet extending `DefaultNonTerminalOpTree`.

exactly! Should I take care of it or do you?

**Comment by jrudolph** [comment]
> exactly! Should I take care of it or do you?

I can do it. I'll also do `quiet`, `atomic`, and `named`.

**Comment by yanns** [comment]
Have you tried scala 3.0.2? 
Lot of errors like:
```
[error]    |Note: a match type could not be fully reduced:
[error]    |
[error]    |  trying to reduce  org.parboiled2.support.TailSwitch.TailSwitch0[org.parboiled2.support.hlist.HNil
[error]    |  ,
[error]    |org.parboiled2.support.hlist.HNil, ?3.Out, ?3.Out, ?4.Out,
[error]    |  org.parboiled2.support.hlist.HNil
[error]    |]
[error]    |  failed since selector  ?3.Out
[error]    |  does not match  case org.parboiled2.support.hlist.HNil => ?4.Out
[error]    |  and cannot be shown to be disjoint from it either.
[error]    |  Therefore, reduction cannot advance to the remaining case
[error]    |
```

**Comment by jrudolph** [comment]
> Have you tried scala 3.0.2?

No, but that's bad news, I guess :) If typing isn't stable, it will be really hard to maintain this stuff...

**Comment by jrudolph** [comment]
@nicolasstucki (or anyone else from the Scala 3 team), the incompatible typing changes in Scala 3.0.2 seem to have messed up the typing here. I unfortunately have no time left to deeper investigate these issues or provide minimal reproducers. If that stays this way it will be a blocker to get akka-http (and the depending Akka ecosystem) to Scala 3. I'll now do an attempt to get akka-http compiling on Scala 3.0.1 but we will have to stay on that version until these problems are solved.

**Comment by jrudolph** [comment]
Also filed as https://github.com/lampepfl/dotty/issues/13491

**Comment by jrudolph** [comment]
From my side, this effort is currently a bit on hold since I currently cannot afford to chase down rabbit hole while some of the typing bits seem still to be in flux in the compiler.

I did some experiments how far I would get integrating the current work into akka-http but I also run into the gnarly typing issues.

That said, I filed a few issues on Scala 3 that got some prompt attention from the team:

 * https://github.com/lampepfl/dotty/issues/13491
 * https://github.com/lampepfl/dotty/issues/13431
 * https://github.com/lampepfl/dotty/issues/13455
 * https://github.com/lampepfl/dotty/issues/13422

Especially, the hanging typer issue (https://github.com/lampepfl/dotty/issues/13431) seems to be solved on master, which would unblock further analysis on what's going on. I would imagine that https://github.com/lampepfl/dotty/issues/13491 (if it can be solved) might get us back on track later.

I'm quite happy with how far we got so quickly. Thanks a lot for the help, @yanns and @lrytz. I hope to spend some time now and then to see what the state is with the latest compiler developments and hopefully we will so arrive at a state that works good enough to be released.

@yanns did you try using the current version with your code base already?

**Comment by odersky** [comment]
> No, but that's bad news, I guess :) If typing isn't stable, it will be really hard to maintain this stuff...

It's really bug fixes. We expect that this would stabilize quickly. We have a lot of projects in the community build that usually port to new versions without problems.

**Comment by jrudolph** [comment]
> > No, but that's bad news, I guess :) If typing isn't stable, it will be really hard to maintain this stuff...
> 
> It's really bug fixes. We expect that this would stabilize quickly. We have a lot of projects in the community build that usually port to new versions without problems.

Yes, I agree. Hopefully, the typing did not rely on to much undefined behavior before, so that fixes will be actually viable.

**Comment by yanns** [comment]
> @yanns did you try using the current version with your code base already?

Not yet, I'm still have some other dependencies to port to scala 3 before... It's a long road for me.

**Comment by odersky** [comment]
Out of interest: Why do you need parboiled2 for akka?

**Comment by jrudolph** [comment]
> Out of interest: Why do you need parboiled2 for akka?

Because akka-http has many HTTP header parsers written defined with parboiled2 and parser execution is somewhat in the hot path. Moving to another parser engine might be an alternative but also requires lots of work to migrate the parsers and we would have to completely reevaluate the performance consequences.

**Comment by sirthias** [comment]
(Akka-http header parsing was indeed the original use case for parboiled2 in the first place. Back in 2013... )

**Comment by odersky** [comment]
> Because akka-http has many HTTP header parsers written defined with parboiled2 and parser execution is somewhat in the hot path. Moving to another parser engine might be an alternative but also requires lots of work to migrate the parsers and we would have to completely reevaluate the performance consequences.

I understand. What I found somewhat strange is that parboiled2 seems to use full-blown typelevel programming with HLists and stuff instead of relying on staging with quotes and splices. What did it do on Scala-2? Use Shapeless HLists?

**Comment by sirthias** [comment]
> I understand. What I found somewhat strange is that parboiled2 seems to use full-blown typelevel programming with HLists and stuff instead of relying on staging with quotes and splices. What did it do on Scala-2? Use Shapeless HLists?

Yes.
We started out with a shapeless dependency and later internalized the HLists (because that was pretty much all we needed from shapeless). So far all type-level logic has been built on these.
[Here](https://github.com/sirthias/parboiled2/blob/6d90b1324c5faa74c38d603e4028afb2dc123372/parboiled-core/src/main/scala/org/parboiled2/support/TailSwitch.scala) is an example.

**Comment by odersky** [comment]
OK, I understand. One thing one could try is to keep HLists as a typeclass the way it is done in Shapeless. Then you'd have implicit search instead of match types. It might be that this is easier to port, but I don't know for sure. The way you did it is more idiomatic for Scala 3, but on the other hand match types are less battle tested than implicit search at present. That said, some of the fine print of implicit search has also changed in Scala 3, so there's no guarantee it would port more easily.

**Comment by jrudolph** [comment]
The regressions have been solved in the latest dotty nightlies, so we are back to a state that worked as in 3.0.1.

https://github.com/sirthias/parboiled2/pull/313

**Comment by xuwei-k** [comment]
- https://github.com/sirthias/parboiled2/pull/324

**Comment by yanns** [comment]
We're missing an implementation for `runSubParser` if anyone wants to pick that.

**Comment by liosedhel** [comment]
Hi ;) First of all, thanks for the effort towards migrating this lib! Is there any time schedule to migrate this lib (and hence akka-http) to Scala 3 in foreseeable future? Maybe there are some areas where one can quite easily contribute?

**Comment by jrudolph** [comment]
> Hi ;) First of all, thanks for the effort towards migrating this lib! Is there any time schedule to migrate this lib (and hence akka-http) to Scala 3 in foreseeable future? Maybe there are some areas where one can quite easily contribute?

The biggest problem is that real world parsers still hang the Scala 3 compiler or fail with typing errors.

We have some tests commented out in `ActionSpec` that show this kind of behavior on a small scale, so you could try to pick one of those tests and see what is needed to make the tests pass. But this is by no means a simple task. This is territory where the old code worked because of un- or underdocumented features of the Scala 2 typing system and we will have to find our ways through to the un- or underdocumented or incompletely implemented features of the Scala 3 compiler.

File: parboiled-core/src/main/scala-3/org/parboiled2/support/TailSwitch.scala [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        case ::[h, t] =>
+          TI match {
+            case HNil      => TailSwitch0[L, t, T, HNil, R, h :: RI]
+            case ::[_, tt] => TailSwitch0[L, t, T, tt, R, h :: RI]
+          }
+      }
+  }
+
+  type Aux[L <: HList, LI <: HList, T <: HList, TI <: HList, R <: HList, RI <: HList, Out <: HList] =
+    TailSwitch[L, T, R] { type Out = TailSwitch0[L, L, T, T, R, HNil] }
```
Comment (prolativ) [inline parboiled-core/src/main/scala-3/org/parboiled2/support/TailSwitch.scala:75]:
```suggestion
  type Aux[L <: HList, LI <: HList, T <: HList, TI <: HList, R <: HList, RI <: HList, Out0 <: HList] =
    TailSwitch[L, T, R] { type Out = Out0 }
```
This seems to fix the remaining problems with `ActionSpec`

File: parboiled-core/src/test/scala/org/parboiled2/ActionSpec.scala [hunk 2]
```diff
@@ -141,10 +147,10 @@ object ActionSpec extends TestParserSpec {
 
       case class Foo(i: Int, s: String)
 
-      "`~>` producing case class (simple notation)" - new TestParser1[Foo] {
+      /*FIXME: type error "`~>` producing case class (simple notation)" - new TestParser1[Foo] {
         def targetRule = rule(push(1 :: "X" :: HNil) ~> Foo)
```
[the remark this answers is not in the document]
Comment (jrudolph) [reply parboiled-core/src/test/scala/org/parboiled2/ActionSpec.scala:151]: Fine for me.

File: parboiled-core/src/main/scala-3/org/parboiled2/support/TailSwitch.scala (hunk 1 above)
Comment (jrudolph) [reply parboiled-core/src/main/scala-3/org/parboiled2/support/TailSwitch.scala:75]: Thanks a lot, sometimes more eyes help a lot spot those errors. Not sure what I was thinking there...

Comment (jrudolph) [reply parboiled-core/src/main/scala-3/org/parboiled2/support/TailSwitch.scala]: Enabling remaining tests in https://github.com/sirthias/parboiled2/pull/349, looking good!

**Comment by liosedhel** [comment]
> We have some tests commented out in `ActionSpec` 
this now seems to be fixed, right? ;) 
thank you @prolativ @KacperFKorban and of course @jrudolph for quick response and looking into it 🙏  Anything else here that one could help with? :)

**Review by jrudolph — COMMENTED** [review]
From my side this now done! Thanks everyone involved. @sirthias / @xuwei-k what would it take to do a new release?

**Comment by sirthias** [comment]
Awesome! Huge thanks again to everyone involved in this massive task!
I'll package and publish a new release right away...

**Comment by sirthias** [comment]
Ok, just released parboiled2 2.4.0, cross-built against:

- 2.12.5
- 2.13.8
- 3.1.1

X

- Java 8
- scala-js 1.9.0

Unfortunately I had to (temporarily) disable the cross-build against _scala-native_ due to a missing dependency (_scala-check_ for _scala-native_ 0.4.3) and weird `binary-incompatible version of NIR` errors, that I didn't have time to debug today.

Thank you again, everyone, for the awesome feat of pulling off this migration!
